datasource/virtual_machine/data.go

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+//go:generate packer-sdc struct-markdown
+//go:generate packer-sdc mapstructure-to-hcl2 -type Config,Tag,DatasourceOutput
+package virtual_machine
+
+import (
+	"github.com/hashicorp/hcl/v2/hcldec"
+	"github.com/hashicorp/packer-plugin-sdk/common"
+	"github.com/hashicorp/packer-plugin-sdk/hcl2helper"
+	packersdk "github.com/hashicorp/packer-plugin-sdk/packer"
+	"github.com/hashicorp/packer-plugin-sdk/template/config"
+	vsCommon "github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common"
+	"github.com/pkg/errors"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// Example of multiple vm_tags blocks in HCL format:
+// ```
+//
+//	vm_tags {
+//	  category = "team"
+//	  name = "operations"
+//	}
+//	vm_tags {
+//	  category = "SLA"
+//	  name = "gold"
+//	}
+//
+// ```
+type Tag struct {
+	// Tag with this name must be attached to virtual machine which should pass the Tags Filter.
+	Name string `mapstructure:"name" required:"true"`
+	// Name of the category that contains this tag. Both tag and category must be specified.
+	Category string `mapstructure:"category" required:"true"`
+}
+
+type Config struct {
+	common.PackerConfig    `mapstructure:",squash"`
+	vsCommon.ConnectConfig `mapstructure:",squash"`
+
+	// Basic filter with glob support (e.g. `nginx_basic*`). Defaults to `*`.
+	// Using strict globs will not reduce execution time because vSphere API returns the full inventory.
+	// But can be used for better readability over regular expressions.
+	Name string `mapstructure:"name"`
+	// Extended name filter with regular expressions support (e.g. `nginx[-_]basic[0-9]*`). Default is empty.
+	// The match of the regular expression is checked by substring. Use `^` and `$` to define a full string.
+	// E.g. the `^[^_]+$` filter will search names without any underscores.
+	// The expression must use [Go Regex Syntax](https://pkg.go.dev/regexp/syntax).
+	NameRegex string `mapstructure:"name_regex"`
+	// Filter to return only objects that are virtual machine templates.
+	// Defaults to `false` and returns all VMs.
+	Template bool `mapstructure:"template"`
+	// Filter to search virtual machines only on the specified node.
+	Node string `mapstructure:"node"`
+	// Filter to return only that virtual machines that have attached all specifies tags.
+	// Specify one or more `vm_tags` blocks to define list of tags that will make up the filter.
+	// Should work since vCenter 6.7. To avoid incompatibility, REST client is being
+	// initialized only when at least one tag has been defined in the config.
+	VmTags []Tag `mapstructure:"vm_tags"`
```

Comment:
Would this be better as just `Tags` / `tags` and the update accordingly? The category defines the associable object type so prepending `vm_` seems irrelevant.